### ISSUE TITLE
Composer generates autoload files on global vendor directory not current working directory.

### ISSUE THREAD

**Issue by MirazMac** [composer/composer#8540]
My `composer.json`:

```json
{
    "autoload": {
        "psr-4": {
            "MirazMac\\MagicMp3\\": "src/"
        }
    }
}
```

Output of `composer diagnose`:

```
Checking composer.json: WARNING
No license specified, it is recommended to do so. For closed-source software you may use "proprietary" as license.
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com rate limit: OK
Checking disk free space: OK
Checking pubkeys:
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.9.2
PHP version: 7.2.9
PHP binary path: H:\wamp64\bin\php\php7.2.9\php.exe
```

When I run this command:

```
composer dump-autoload
```

I get the following output:

```
Generated autoload files containing 0 classes
```

And I expected this to happen:

But there are classes inside the **src** directory, thus autoload files should be generated. 

Oddly, when I run this command:

```
composer dump-autoload -o
```
I get the following output:
```
Generated optimized autoload files containing 5 classes
```
And I expected this to happen:

But the problem is, this time, the autoload files are generated inside the global vendor directory (**C:\Users\MirazMac-Lapi\AppData\Roaming\Composer\vendor**).

But no vendor directory is generated from where I ran the command, the current working directory, where my `composer.json` is located.

This issue is not only occurring with `composer dump-autoload` but also with any command that generates autoload files. I'm facing the same issues with `composer install` , `composer require` and `composer update`.

I tried a complete uninstall and then re-installing the latest version of Composer. But it didn't help. Then again I tried switching between git bash and Windows CMD. And I made sure that I'm indeed on the correct working directory. But I'm lost at this point.

**Comment by MirazMac** [comment]
Hello.. Anyone?

**Comment by Seldaek** [comment]
We'd need to see the full path name and the contents (at least namespace + class name) of those classes you expect it to find.. seems like a misconfiguration.

**Comment by MirazMac** [comment]
> We'd need to see the full path name and the contents (at least namespace + class name) of those classes you expect it to find.. seems like a misconfiguration.

It's not happening only for this project, basically any composer require/install/update command is installing the packages on the global directory! I even tried force setting the working-dir and still it installs and generates autoload files on the global directory. I tried complete un-install and re-install several times, with no luck.

**Comment by Seldaek** [comment]
In that case please send the output of a `composer dump-autoload -o -vvv` run, maybe that'll highlight an issue.. And include the full path you ran the command in as well.

**Comment by MirazMac** [comment]
Okay, so I ran the command: `composer dump-autoload -o -vvv` from here:
`H:\wamp64\www\mp3magic`

Here's the output:
`
Reading ./composer.json
Loading config file ./composer.json
Checked CA file C:\Users\MirazMac-Lapi\AppData\Local\Temp\opeD2B1.tmp: valid
Executing command (H:\wamp64\www\mp3magic): git branch --no-color --no-abbrev -v
Executing command (H:\wamp64\www\mp3magic): git describe --exact-match --tags
Executing command (H:\wamp64\www\mp3magic): git log --pretty="%H" -n1 HEAD
Executing command (H:\wamp64\www\mp3magic): hg branch
Executing command (H:\wamp64\www\mp3magic): fossil branch list
Executing command (H:\wamp64\www\mp3magic): fossil tag list
Executing command (H:\wamp64\www\mp3magic): svn info --xml
Reading C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Loading config file C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Running 1.9.2 (2020-01-14 16:30:31) with PHP 7.4.2 on Windows NT / 10.0
Generated optimized autoload files containing 8 classes
`

But no vendor directory is generated on the directory from where I run the command, instead the autoload files are generated once again, on the global vendor directory.

**Comment by Seldaek** [comment]
Well I can't reproduce this.. But what I find strange is that the `Running 1.9.2 ...` line is the first line for me, and not for you. It may be you have a plugin or something doing weird stuff. Can you try with `--no-plugins` added to it? Also maybe check what `composer` actually runs.. try `where composer` and then see what that is, it's maybe doing something strange?

**Comment by MirazMac** [comment]
Ah, sorry but it's a fresh install and no plugin is installed. And running `where composer` shows the correct location:
```
C:\ProgramData\ComposerSetup\bin\composer
C:\ProgramData\ComposerSetup\bin\composer.bat
```
I guess I can't use composer then for a while. I have never faced such issue in my life.

**Comment by Seldaek** [comment]
Plugins can be installed globally, what does `composer global show` look like?

**Comment by MirazMac** [comment]
No what I mean was it is a fresh installation of composer with no plugins installed, `composer global show` outputs 
```
Changed current directory to C:/Users/MirazMac-Lapi/AppData/Roaming/Composer
```

and these are the only files the directory has, hint: default files installed by composer.

```
│   .htaccess
│   cacert.pem
│   composer.json
│   keys.dev.pub
│   keys.tags.pub
│
└───vendor
    │   autoload.php
    │
    └───composer
            autoload_classmap.php
            autoload_namespaces.php
            autoload_psr4.php
            autoload_real.php
            autoload_static.php
            ClassLoader.php
            LICENSE
```

**Comment by Seldaek** [comment]
Yeah sorry for double-checking you but I'm trying to exclude any possible problem I can think of here.. 

Have you tried with older versions? You can run `composer self-update 1.6.0` for example to downgrade it, if you can identify the point where it started breaking that could give us a hint..

**Comment by MirazMac** [comment]
Okay so I've downgraded to 1.6.0 and and still when I run `composer require mirazmac/bangla-string -vvv` on an empty directory it generates the `composer.json` and `composer.lock` file on that directory but the package is still being installed on the global vendor directory.

Output for the require command:

```
Reading C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Loading config file C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Checked CA file C:\Users\MirazMac-Lapi\AppData\Local\Temp\ope5581.tmp: valid
Executing command (C:/Users/MirazMac-Lapi/AppData/Roaming/Composer): git branch --no-color --no-abbrev -v
Executing command (C:/Users/MirazMac-Lapi/AppData/Roaming/Composer): git describe --exact-match --tags
Executing command (C:/Users/MirazMac-Lapi/AppData/Roaming/Composer): git log --pretty="%H" -n1 HEAD
Executing command (C:/Users/MirazMac-Lapi/AppData/Roaming/Composer): hg branch
Executing command (C:/Users/MirazMac-Lapi/AppData/Roaming/Composer): fossil branch list
Executing command (C:/Users/MirazMac-Lapi/AppData/Roaming/Composer): fossil tag list
Executing command (C:/Users/MirazMac-Lapi/AppData/Roaming/Composer): svn info --xml
Reading C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Loading config file C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Running 1.6.0 (2018-01-04 12:03:22) with PHP 7.4.2 on Windows NT / 10.0
Reading ./composer.json
Loading config file ./composer.json
Executing command (C:\Users\MirazMac-Lapi\Desktop): git branch --no-color --no-abbrev -v
Executing command (C:\Users\MirazMac-Lapi\Desktop): git describe --exact-match --tags
Executing command (C:\Users\MirazMac-Lapi\Desktop): git log --pretty="%H" -n1 HEAD
Executing command (C:\Users\MirazMac-Lapi\Desktop): hg branch
Executing command (C:\Users\MirazMac-Lapi\Desktop): fossil branch list
Executing command (C:\Users\MirazMac-Lapi\Desktop): fossil tag list
Executing command (C:\Users\MirazMac-Lapi\Desktop): svn info --xml
Reading C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Loading config file C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Downloading https://packagist.org/packages.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/packages.json into cache
Downloading http://packagist.org/p/provider-2013%24b42e10bb8c0683f600dc825c8331e5e1b910acf0de4733c09879bfc590b8b4f1.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2013.json into cache
Downloading http://packagist.org/p/provider-2014%2491f4bf506e01bb4413bf5c8b382fe0f8798e9b1009362443dd551dbdda23fde3.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2014.json into cache
Downloading http://packagist.org/p/provider-2015%2424b1e51a03ba26d11c22a65895f50230b431cb760d5737d99ffb1258ed71fd17.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2015.json into cache
Downloading http://packagist.org/p/provider-2016%24da85204da998c57690d51a8efef67714c75be5a6725de45c2dbeb13b6e4638a5.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2016.json into cache
Downloading http://packagist.org/p/provider-2017%246176f0a07fdcfd1c9feae84adc9cd16c77b68c3d00f7e756642aa716c98ecdd8.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2017.json into cache
Downloading http://packagist.org/p/provider-2018%2461661615b4f19d8353b0d05208bfe4012c2da416d6cd8a85cc6eb52ffb3b799b.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2018.json into cache
Downloading http://packagist.org/p/provider-2019%24897bef3ae1a6a48c47ddf32daf94429ddbbb484566f6c91cc29e96cf5cb2655c.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2019.json into cache
Downloading http://packagist.org/p/provider-2019-04%247281e61bbbee018f9dc02bced1a2b9a5890e4b83657e5cbb945bb54dd7db453a.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2019-04.json into cache
Downloading http://packagist.org/p/provider-2019-07%249b6ddcbdf58c4e9033c8d4b9b4b44c079b073b10f1e0d31e034723c041a67ee6.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2019-07.json into cache
Downloading http://packagist.org/p/provider-2019-10%240615f7d095d5b92e21381d49e1d723fd5e65472020e4dddcfc43edca031e45ad.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2019-10.json into cache
Downloading http://packagist.org/p/provider-2020-01%24af7b86d1f014ddcd3475729723cf007217c9ce9eb7e1928107b7ce6897b6f437.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2020-01.json into cache
Downloading http://packagist.org/p/provider-archived%243659e1651fa6a3cb79ab932b4c5848a14485d448e4a201cfc1527b24df3938e3.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-archived.json into cache
Downloading http://packagist.org/p/provider-latest%24e97422bd11052a3a08f8275abc4561cf0335c841fe2384697d12def52639fa8c.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-latest.json into cache
Downloading http://packagist.org/p/mirazmac/bangla-string%24a77ffbb86c9b4c103926bafe7c8eb3918d3fa42fd3bfb1778e409b93fa9f8ecd.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/provider-mirazmac$bangla-string.json into cache
Using version ^1.1 for mirazmac/bangla-string
./composer.json has been created
Reading ./composer.json
Loading config file ./composer.json
Executing command (C:\Users\MirazMac-Lapi\Desktop): git branch --no-color --no-abbrev -v
Executing command (C:\Users\MirazMac-Lapi\Desktop): git describe --exact-match --tags
Executing command (C:\Users\MirazMac-Lapi\Desktop): git log --pretty="%H" -n1 HEAD
Executing command (C:\Users\MirazMac-Lapi\Desktop): hg branch
Executing command (C:\Users\MirazMac-Lapi\Desktop): fossil branch list
Executing command (C:\Users\MirazMac-Lapi\Desktop): fossil tag list
Executing command (C:\Users\MirazMac-Lapi\Desktop): svn info --xml
Reading C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Loading config file C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/composer.json
Loading composer repositories with package information
Downloading https://packagist.org/packages.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/packages.json into cache
Updating dependencies (including require-dev)
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2013.json from cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2014.json from cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2015.json from cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2016.json from cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2017.json from cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2018.json from cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2019.json from cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2019-04.json from cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2019-07.json from cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2019-10.json from cache
Downloading http://packagist.org/p/provider-2020-01%24b6ad49945259b33b197e68dbc0a340624675202165cdfa93dfd867025f812f48.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-2020-01.json into cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-archived.json from cache
Downloading http://packagist.org/p/provider-latest%2471ab3974fd4ab6d59228a98a9ad8d7901bb88b411a5a4b2449ddb7c68914c238.json
Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/p-provider-latest.json into cache
Reading C:/Users/MirazMac-Lapi/AppData/Local/Composer/repo/https---packagist.org/provider-mirazmac$bangla-string.json from cache
Resolving dependencies through SAT
Dependency resolution completed in 0.000 seconds
Analyzed 60 packages to resolve dependencies
Analyzed 59 rules to resolve dependencies
Package operations: 1 install, 0 updates, 0 removals
Installs: mirazmac/bangla-string:v1.1
  - Installing mirazmac/bangla-string (v1.1): Downloading https://api.github.com/repos/MirazMac/BanglaString/zipball/dcf106bc5ccfaf9d1d2ba6cbf3f3331428e75491
Downloading (connecting...)
Following redirect (2) https://codeload.github.com/MirazMac/BanglaString/legacy.zip/dcf106bc5ccfaf9d1d2ba6cbf3f3331428e75491
Downloading https://codeload.github.com/MirazMac/BanglaString/legacy.zip/dcf106bc5ccfaf9d1d2ba6cbf3f3331428e75491
Downloading (100%)                                                               Writing C:/Users/MirazMac-Lapi/AppData/Local/Composer/files/mirazmac/bangla-string/c202d86d1e9aa17aecd5dbabb8e8ebf8476d05e9.zip into cache from C:\Users\MirazMac-Lapi\AppData\Roaming\Composer\vendor/mirazmac/bangla-string/3fbeb19e925ca909d5df91e992b87516

 Extracting archive    REASON: Required by the root package: Install command rule (install mirazmac/bangla-string v1.1)

Reading C:\Users\MirazMac-Lapi\AppData\Roaming\Composer\vendor/composer/installed.json
Writing lock file
Generating autoload files
```

**Comment by Seldaek** [comment]
Can you try even older versions from https://getcomposer.org/download/#composer-history-caption ? If they all fail then maybe you'll need to clone the project and run it from source so you can see what is going on..

**Comment by MirazMac** [comment]
Thanks for the tip, I'll try and let you know.

**Comment by Seldaek** [comment]
Also can you share the contents of C:\ProgramData\ComposerSetup\bin\composer and 
C:\ProgramData\ComposerSetup\bin\composer.bat ?

It seems to me like you are running the composer global command somehow, at least it behaves in a similar way, but not completely. It's really weird.

**Comment by MirazMac** [comment]
Contents of: **C:\ProgramData\ComposerSetup\bin\composer**
```
#!/bin/sh

dir=$(cd "${0%[/\\]*}" > /dev/null; pwd)

if [ -d /proc/cygdrive ]; then
    case $(which php) in
        $(readlink -n /proc/cygdrive)/*)
            # We are in Cygwin using Windows php, so the path must be translated
            dir=$(cygpath -m "$dir");
            ;;
    esac
fi

php "${dir}/composer.phar" "$@"
```

Contents of the composer.bat file:

```
@echo OFF
:: in case DelayedExpansion is on and a path contains ! 
setlocal DISABLEDELAYEDEXPANSION
php "%~dp0composer.phar" %*
```

**Comment by stof** [comment]
is `C:/Users/MirazMac-Lapi/AppData/Roaming/Composer/` your working directory ?

If no, something looks weird in the output of `composer require`, as that's where it works.
Do you have a `COMPOSER` env variable set ?

**Comment by MirazMac** [comment]
No, that's not my working directory. And yes that is what I'm talking about, it's super weird. Like in my working directory composer is updating/generating the `composer.json` and `composer.lock` file but God knows why the packages and autoload files are being installed on the global directory.

No I don't have the COMPOSER env variable set.

**Comment by MirazMac** [comment]
And look at the contents of the `C:\Users\MirazMac-Lapi\AppData\Roaming\Composer\vendor\composer\autoload_psr4.php` file:

```
<?php

// autoload_psr4.php @generated by Composer

$vendorDir = dirname(dirname(__FILE__));
$baseDir = 'h:/wamp64/www/mp3magic';

return array(
    'MirazMac\\MagicMp3\\' => array($baseDir . '/src'),
);
```

Here the base directory is set to my current working directory but the autoload files themselves are on the global directory!

**Comment by Seldaek** [comment]
Having COMPOSER_VENDOR_DIR set would possibly do this, can you check with `echo %COMPOSER_VENDOR_DIR%` that it is not set?

**Comment by MirazMac** [comment]
I don't have it set.

`echo %COMPOSER_VENDOR_DIR%`
output:
```
%COMPOSER_VENDOR_DIR%
```

But then I did this, I uninstalled composer, and this time before installing it again, I performed a restart and this time I installed composer via the manual commands from the getcomposer.org page not using the windows installer. And it seems to work fine.

**Comment by Seldaek** [comment]
Ok now this is even weirder.. Looping in @johnstevenson who maybe has a clue why the exe installer would trigger such madness :)

**Comment by MirazMac** [comment]
Never mind, I guess it's not an issue of the installer. The thing just happened again, and now that I check manually, %COMPOSER_VENDOR_DIR% does have a value. But oddly from CMD it's only echoing if the CMD is in administrator mode (run as administrator).

So, I manually removed the variable from the entry from the environment variables. And ran `refreshenv` 
and seems like everything's back to normal.
The real weird thing is that how that variable got set in the first place. I never knew such env var existed unless you told me to check for it. And thank God you mentioned it otherwise I was going to re-install Windows and everything. 
Thanks a lot for the support man. Appreciate it.

**Comment by Seldaek** [comment]
Huh, ok, curious to hear if @johnstevenson has a clue still, if perhaps the env var could be set by the installer. Otherwise no clue how this could've happened except if you're sleepwalking or something ;) Anyway good thing it's fixed.

**Comment by MirazMac** [comment]
Haha, I won't consider sleep walking a possible issue with me, although I wouldn't know cause I'd be sleeping :3

But my guess is some scripts (bash or batch) added that env variable, because the first time I remember this happening I was really bothered, because it was working fine the day before that. And all I did was using composer regularly (installing, updating). Anyway, I'm gonna do a sweep and see if I can find the culprit or not, I'll let you guys know. Thanks once again.

**Comment by johnstevenson** [comment]
> Never mind, I guess it's not an issue of the installer. The thing just happened again, and now that I check manually, %COMPOSER_VENDOR_DIR% does have a value. But oddly from CMD it's only echoing if the CMD is in administrator mode (run as administrator).

Composer-Setup.exe has never set `%COMPOSER_VENDOR_DIR%`. What makes no sense is that you see a different environment in cmd depending whether it is elevated or not.

Are you launching cmd via some wrapper program (ConEmu, Commander etc)? Alternatively you might have something in the registry (under the `HKLM/HKCU Software\Microsoft\Command Processor\AutoRun` key)

**Comment by MirazMac** [comment]
Yeah, I use Git Bash mostly. But I did try once or twice using CMD.

I understand that Composer-Setup has nothing to do with it. Because the first time this occurred was out of the blue. As I used the installer to setup Composer a long time ago and everything worked fine since ages. And suddenly this happened. 

I'm assuming some script or program messed with the environment variables. Could be a misconfiguration issue. idk at this point I'm just happy that it's working.

I'll check if I can find what caused the env variable to set but I'm sure now it wasn't Composer.
